uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-calendar-day.component.js

```diff
@@ -168,44 +98,130 @@ define([
-        ? 'accrued-toil-tooltip'
-        : getTooltipTemplateForLeaveRequests();
+    function resolveLeaveRequestsCalculationUnits () {
+      var leaveRequestAttributes, absenceType, calculationUnit;
+
+      vm.contactData.leaveRequests.forEach(function (leaveRequest) {
+        leaveRequestAttributes = vm.contactData.leaveRequestsAttributes[leaveRequest.id];
+        absenceType = _.find(vm.supportData.absenceTypes, { id: leaveRequest.type_id });
+        calculationUnit = _.find(vm.supportData.calculationUnits, { 'value': absenceType.calculation_unit });
+        leaveRequestAttributes.unit = calculationUnit.name;
+      });
+    }
+
+    /**
+     * Sets dates from leave requests to leave requests attributes
+     * by converting them from String to Date type.
+     */
+    function resolveLeaveRequestsDates () {
+      var leaveRequestAttributes;
+
+      vm.contactData.leaveRequests.forEach(function (leaveRequest) {
+        leaveRequestAttributes = vm.contactData.leaveRequestsAttributes[leaveRequest.id];
+        leaveRequestAttributes.from_date =
+          new Date(leaveRequest.from_date);
+        leaveRequestAttributes.to_date =
+          new Date(leaveRequest.to_date);
+      });
+    }
+
+    /**
+     * Sets the from and to date type labels to the leave requests attributes.
+     */
+    function resolveLeaveRequestsTypes () {
+      var leaveRequestAttributes, fromDateType, toDateType;
+
+      vm.contactData.leaveRequests.forEach(function (leaveRequest) {
+        leaveRequestAttributes = vm.contactData.leaveRequestsAttributes[leaveRequest.id];
+
+        if (leaveRequestAttributes.unit === 'days') {
+          fromDateType = findRecordByIdFieldValue(vm.supportData.dayTypes,
+            'value', leaveRequest.from_date_type);
+          toDateType = findRecordByIdFieldValue(vm.supportData.dayTypes,
+            'value', leaveRequest.to_date_type);
+
+          leaveRequestAttributes['from_date_type'] = fromDateType.label;
+          leaveRequestAttributes['to_date_type'] = toDateType.label;
+        }
+      });
     }
 
     /**
      * Waits for the leave request to be accesible before mapping the necessary
      * leave request fields to it.
      */
     function watchForLeaveRequestReady () {
-      $scope.$watch('day.contactData.leaveRequest', function () {
-        if (vm.contactData && vm.contactData.leaveRequest) {
-          mapLeaveRequestAbsenceType();
-          findAbsenceTypeCalculationUnit();
-          mapLeaveRequestDateTypes();
-          resolveDayLabel();
-          resolveDayDates();
-          selectTooltipTemplate();
+      $scope.$watch('day.contactData.leaveRequests', function () {
+        if (vm.contactData && vm.contactData.leaveRequests) {
+          resolveLeaveRequestsCalculationUnits();
+          resolveLeaveRequestsTypes();
+          resolveLeaveRequestsLabels();
+          resolveLeaveRequestsDates();
+          resolveLeaveRequestsAbsenceTypesTitles();
```

Comment:
I notice that we are repeating the same pattern in each one of these functions:

* Cycle through each one of `vm.contactData.leaveRequests`.
* Find the corresponding `leaveRequestAttributes`.
* resolve the particular data that you are trying to find.

Why don't we do something a bit different. Is it ok to hold all the information in a different array of objects?

```js
$scope.$watch('day.contactData.leaveRequests', function () {
  if (vm.contactData &&  vm.contactData.leaveRequests) {
    vm.leaveRequestEntries = resolveLeaveRequestEntries(); // or a better name please
  }
});

function resolveLeaveRequestEntries () {
  return vm.contactData.leaveRequests.map(function (leaveRequest) {
    var attributes = _.cloneDeep(vm.contactData.leaveRequestsAttributes[leaveRequest.id]);
    
    attributes.dateTypes = resolveLeaveRequestsDateTypes(leaveRequest, attributes);
    attributes.label = resolveLeaveRequestLabel(leaveRequest, attributes);
    attributes.from_date = new Date(leaveRequest.from_date);
    attributes.to_date = new Date(leaveRequest.to_date);
    attributes.absenceTypeTitle = getLeaveRequestAbsenceType(leaveRequest).title;

    return {
      attributes: attributes,
      leaveRequest: leaveRequest
    };
  });
}
```

I think this approach would remove the number of functions that you'll need and the process seems more streamlined. You'd need to use `vm.leaveRequestEntries` on the view though.